packages/markers/src/browser/markers-tree.ts

```diff
@@ -47,53 +44,55 @@ export class MarkersTree extends Tree {
     }
 
     getMarkerInfoNodes(parent: MarkerRootNode): Promise<MarkerInfoNode[]> {
-        const markerFiles = this.markersManager.getMarkerInformationByKind(this.markerRoot.kind);
         const uriNodes: MarkerInfoNode[] = [];
-        markerFiles.forEach(markerInfo => {
-            const cachedMarkerInfo = this.markerInfoNodeCache.find(m => m.uri.toString() === markerInfo.uri);
-            const markerInfoNodeUri = new URI(markerInfo.uri);
-            const markerInfoNode: MarkerInfoNode = {
-                children: [],
-                expanded: cachedMarkerInfo ? cachedMarkerInfo.expanded : false,
-                uri: markerInfoNodeUri,
-                id: 'markerInfo-' + markerInfoNodeUri.displayName,
-                name: markerInfoNodeUri.displayName,
-                parent,
-                selected: false,
-                numberOfMarkers: markerInfo.counter
-            };
-            uriNodes.push(markerInfoNode);
+        this.markersManager.forEachMarkerInfoByKind(this.markerRoot.kind, markerInfo => {
+            const uri = new URI(markerInfo.uri);
+            const id = 'markerInfo-' + markerInfo.uri;
+            const cachedMarkerInfo = this.getNode(id);
+            if (cachedMarkerInfo && MarkerInfoNode.is(cachedMarkerInfo)) {
+                uriNodes.push(cachedMarkerInfo);
+            } else {
+                uriNodes.push({
+                    children: [],
+                    expanded: false,
+                    uri,
+                    id,
+                    name: uri.displayName,
+                    parent,
+                    selected: false,
+                    numberOfMarkers: markerInfo.counter
+                });
+            }
         });
-        this.markerInfoNodeCache = uriNodes;
         return Promise.resolve(uriNodes);
     }
 
     getMarkerNodes(parent: MarkerInfoNode): Promise<MarkerNode[]> {
-        const markers = this.markersManager.getMarkersByUriAndKind(parent.uri.toString(), parent.parent.kind);
         const markerNodes: MarkerNode[] = [];
         let counter: number = 0;
-        this.getNode()
-        markers.forEach((marker: Marker) => {
+        this.markersManager.forEachMarkerByUriAndKind(parent.uri.toString(), parent.parent.kind, marker => {
             counter++;
-            const cachedMarkerNode = this.markerNodeCache.find(m => m.uri.toString() === marker.uri);
             const uri = new URI(marker.uri);
-            const markerNode: MarkerNode = {
-                id: 'marker-' + uri.toString() + '-' + counter,
-                name: marker.kind,
-                parent,
-                selected: cachedMarkerNode ? cachedMarkerNode.selected : false,
-                uri,
-                marker
-            };
-            markerNodes.push(markerNode);
+            const cachedMarkerNode = this.getNode(marker.id);
+            if (MarkerNode.is(cachedMarkerNode)) {
+                markerNodes.push(cachedMarkerNode);
```

Comment:
Should not you reset the marker? e.g. if an error becomes a warning.

Reply:
I am not sure if this is too simple:
https://github.com/theia-ide/theia/commit/ac0e1e94cab4a13848f308ad6c3a9374341cb518#diff-b17954869c4dc749299bb098b8a07c92R79

What would you say?

Reply:
It is simple like that :)